### PR TITLE
Add Puma C extension patches

### PR DESCRIPTION
**Pull request by bjfish**
These patches are derived from the changes found here: https://github.com/eregon/puma/tree/truffle

### CODE REVIEW

**Comment by bjfish** [comment]
@eregon I did not include your `rb` file change here. I imagine that could be done with our ruby patching system.

File: lib/cext/preprocess.rb [hunk 1]
```diff
@@ -49,7 +62,35 @@
                  'websocket_mask.c'    => {:gem   => 'websocket-driver',
                                            :patches => 
                                              [{:match       => /(VALUE .*?)\s+= Qnil;/,
-                                               :replacement => '\1;'}]}}
+                                               :replacement => '\1;'}]},
+                 'http11_parser.c'    => {:gem   => 'puma_http11',
+                                           :patches => 
+                                             [{:match       => /parser->(.*?) = Qnil/,
```
Comment (eregon) [inline lib/cext/preprocess.rb]: I would use `(\w+)` to avoid matching arbitrary complex expressions.

Comment (eregon) [inline lib/cext/preprocess.rb]: And either `$` at the end or at least `;` so `Qnilfoobar` doesn't match :)

File: lib/cext/preprocess.rb [hunk 2]
```diff
@@ -49,7 +62,35 @@
...
+                                               :replacement => 'parser->\1 = rb_tr_handle_for_managed(Qnil)'}]},
+                 'puma_http11.c'      => {:gem   => 'puma_http11',
+                                           :patches => 
+                                             [{:match       => /(define.*?,) Qnil }/,
```
Comment (eregon) [inline lib/cext/preprocess.rb]: I would match those more specifically, as they could be valid if used in a function and not just to initialize a global variable.
```c
# define f(N) { (sizeof(N) - 1), N, 0, Qnil }
# define fr(N) { (sizeof(N) - 1), N, 1, Qnil }
```

File: lib/cext/preprocess.rb [hunk 3]
```diff
@@ -49,7 +62,35 @@
...
+                                               :replacement => '\1 NULL }'},
+                                               {:match       => /cf->value = (.*?);/,
+                                               :replacement => 'cf->value = rb_tr_handle_for_managed(\1);'},
+                                               {:match       => /return found \? found->value : Qnil;/,
+                                               :replacement => 'return found ? rb_tr_managed_from_handle(found->value) : Qnil;'},
+                                               {:match       => /return cf->value;/,
+                                               :replacement => 'return rb_tr_managed_from_handle(cf->value);'},
+                                               {:match       => /void HttpParser_mark\(puma_parser\* hp\) {.*?}/m,
+                                               :replacement => 'void HttpParser_mark(puma_parser* hp) { }'},
```
Comment (eregon) [inline lib/cext/preprocess.rb]: This patch shouldn't be needed, I did it for clarity but we don't call mark functions anyway and `rb_gc_mark` is noop.

File: lib/cext/preprocess.rb [hunk 4]
```diff
@@ -49,7 +62,35 @@
...
+                                               {:match       => /void HttpParser_free\(void \*data\) {.*?}.*?}/m,
```
Comment (eregon) [inline lib/cext/preprocess.rb]: We could match the end of the function with `^\}` if we assume indentation is correct.

File: lib/cext/preprocess.rb [hunk 5]
```diff
@@ -49,7 +62,35 @@
...
+                                               :replacement => PUMA_HTTP_PARSER_FREE},
+                                               {:match       => /\(hp->request/,
```
Comment (eregon) [inline lib/cext/preprocess.rb]: Bets to always match on both extremes, here with `\b` at the end.

**Comment by eregon** [comment]
The Ruby diffs are just for debugging and building the gem from the source repo, they don't apply here indeed.

File: lib/cext/preprocess.rb (hunk 2 above)
Comment (eregon) [inline lib/cext/preprocess.rb]: I missed the `,` which indeed is a good hint that we need it.

**Comment by eregon** [comment]
With this, it seems Puma actually works on a Hello World example!
```ruby
# app.ru
require 'rack'

app = Proc.new do |_env|
  ['200', {'Content-Type' => 'text/html'}, ['A barebones rack app.']]
end

run app
```
```bash
jt ruby -S puma app.ru
curl -v 'http://localhost:9292'
```

Via rackup, we currently fail though:
```bash
jt ruby -S rackup -s puma app.ru
Puma starting in single mode...
* Version 3.10.0 (truffleruby 0.0 - ruby 2.3.3), codename: Russell's Teapot
* Min threads: 0, max threads: 16
* Environment: development
* Listening on tcp://localhost:9292
src/main/ruby/core/library.rb:57:in `getifaddrs': FFI::Library method getifaddrs with caller lib/rubysl/rubysl-socket/lib/rubysl/socket/foreign.rb not implemented (RuntimeError)
	from lib/rubysl/rubysl-socket/lib/socket/socket.rb:178:in `getifaddrs'
	from lib/rubysl/rubysl-socket/lib/socket/socket.rb:5:in `ip_address_list'
	from lib/ruby/gems/2.3.0/gems/puma-3.10.0/lib/puma/binder.rb:249:in `loopback_addresses'

```

**Pull request merged**